pandas_gbq/gbq.py

```diff
@@ -761,7 +761,7 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
     ----------
     query : str
         SQL-Like Query to return data values
-    project_id : str
+    project_id : str (optional)
```

Comment:
The `project_id` is still required when using user authentication, so we should clarify that it is only optional if using default credentials.

Or maybe also optional with a service account? I think we might have to parse as JSON and grab the project ID from the service account and use https://google-auth.readthedocs.io/en/latest/reference/google.oauth2.service_account.html#google.oauth2.service_account.Credentials.from_service_account_info for that to work.

Reply:
Thanks for bearing with me on the authentication questions

> The project_id is still required when using user authentication

Is this right? If I'm running localling with user auth, `google.auth` can pick up my default project fine:

```
In [20]: import google.cloud.bigquery

In [21]: c=google.cloud.bigquery.Client()

In [22]: c
Out[22]: <google.cloud.bigquery.client.Client at 0x1138e44a8>

In [23]: c.project
Out[23]: 'sixty-capital'
```

Auth:
```
gcloud auth list                                             
       Credentialed Accounts
ACTIVE  ACCOUNT
*       []@sixtycapital.com

To set the active account, run:
    $ gcloud config set account `ACCOUNT`
```

Reply:
`google.cloud.bigquery.Client()` does user auth using the Cloud SDK. I'm referring to when the Cloud SDK is not installed, where pandas-gbq does it's user auth flow.

Reply:
To be more specific the credential creation at

https://github.com/pydata/pandas-gbq/blob/16747609dc58f0b6df2c4e896db5a3f365c4677c/pandas_gbq/gbq.py#L272-L283

does not include a default project.

Reply:
Bear with me once more:
Why don't we add it? It's in the JSON file:

![image](https://user-images.githubusercontent.com/5635139/37006348-f11aa8be-20a5-11e8-8dd6-7a54736ef54c.png)

Reply:
In 3LO there are two projects we are concerned with:

- The client secrets project. In this case, it is the project for Pandas GBQ. It has the BigQuery API enabled, but users shouldn't have permissions to run queries against this project. If we did use it, it would bill the Pandas team, not the user.
- The user's project. We know this in the case of a service account or when using the `gcloud` command, but we don't know it if using client secrets.

Reply:
OK I see now. Because we do the `app_flow`, those creds don't come with a project.
Though not sure what 3LO is

Reply:
3LO - 3-legged OAuth 2. (The protocol that `app_flow` follows.)